### PR TITLE
Message formater improvements

### PR DESCRIPTION
**Pull request by benallard**
I started reworking the doc as well, but stopped midway. It's a bigger work as I thought.

### CODE REVIEW

**Comment by benallard** [comment]
... and rebased

**Comment by benallard** [comment]
A big drawback of that serie, is that it takes the template once, and never refresh it ...

**Comment by tardyp** [comment]
> A big drawback of that serie, is that it takes the template once, and never refresh it ...

I think that is cool and more optimized (template parsing only happens once)

We need to make sure that buildbot reconfig properly refresh the template however.
(this means that we have to play with the comparablemixin so that if the loaded template is different, it will see that and switch the reporter's config)

**Comment by benallard** [comment]
The refreshing has to happens inside the reporter then ?

**Comment by tardyp** [comment]
@benallard indeed, the reporter is the BuildbotService here that is reconfigured.
BuildbotService are only reconfigured if the new configs service with the same name is found to be different as per python comparaison.
We use ComparableMixin in order to describe what are the criteria for two objects of the same class to be found different (the default == will else always say it is different)

That said, I can see in the code that mail reporter is not using ComparableMixin, so it should always reconfigure the mail reporter. Could you please make sure of it?

**Comment by benallard** [comment]
I've been running reconfig the whole day, and it refreshed the template.

**Comment by tardyp** [comment]
@benallard there are still some lint issues reported by the CI

File: master/buildbot/reporters/message.py [hunk 1]
```diff
@@ -16,21 +32,48 @@ class MessageFormatter(object):
     template_type = 'plain'
     wantProperties = True
     wantSteps = False
+    wantLogs = False
 
-    def __init__(self, template_name=None, template_dir=None, template_type=None):
+    def __init__(self, template_name=None, template_dir=None, template=None,
+                 subject_name=None, subject=None, template_type=None, ctx=None):
```
Comment (tardyp) [inline master/buildbot/reporters/message.py]: subject_name is weird. probably this is a good time to change to template_filename and subject_filename (with a deprecation warning on template_name)

File: master/buildbot/reporters/message.py [hunk 2]
```diff
@@ -109,13 +152,12 @@ def messageSummary(self, build, results):
         return text
 
     def __call__(self, mode, buildername, buildset, build, master, previous_results, blamelist):
-        """Generate a buildbot mail message and return a tuple of message text
-            and type."""
+        """Generate a buildbot mail message and return a dictionnary
```
Comment (tardyp) [inline master/buildbot/reporters/message.py]: dictioNary

File: master/docs/manual/cfg-reporters.rst [hunk 3]
```diff
@@ -109,137 +109,38 @@ If you want to require Transport Layer Security (TLS), then you can also set ``u
...
-                'body': u"\n".join(text),
-                'type': 'html'
-            }
+                                messageFormatter=reporters.MessageFormatter(template="STATUS: {{ summary }}"))
+
+Another example of a function delivering a customized html email is given below::
+
+    from buildbot.plugins import reporters
+
+    # XXX: This is work in progress, use with care.
```
Comment (tardyp) [inline master/docs/manual/cfg-reporters.rst]: I think this is cool example. No need for warning

File: master/docs/manual/cfg-reporters.rst [hunk 4]
```diff
@@ -109,137 +109,38 @@ If you want to require Transport Layer Security (TLS), then you can also set ``u
...
+    '''
 
     mn = reporters.MailNotifier(fromaddr="buildbot@example.org",
                                 sendToInterestedUsers=False,
                                 mode=('failing',),
                                 extraRecipients=['listaddr@example.org'],
-                                messageFormatter=html_message_formatter)
+                                messageFormatter=reporters.MessageFormatter(
+                                    template=template, template_type='html',
+                                    wantProperties=True, wantSteps=True, wantLogs=True))
```
Comment (tardyp) [inline master/docs/manual/cfg-reporters.rst]: don't put wantLogs by default. its not needed by this template, and its really a lot of overhead

Comment (tardyp) [inline master/docs/manual/cfg-reporters.rst]: I think while we are at it we should really document that MessageFormatter, describing all the available parameters and their implication.
The more details you ask via wantXXX the more overhead the mail reporter will have (memory and CPU)
There is a completely obsolete documentation of messageFormater later in this file

**Comment by benallard** [comment]
CI is happy :)

**Comment by tardyp** [comment]
Cool!

Le lun. 24 oct. 2016 16:18, Benoît Allard <notifications@github.com> a
écrit :

**Pull request merged**